### PR TITLE
8335664: Parsing jsr broken: assert(bci>= 0 && bci < c->method()->code_size()) failed: index out of bounds

### PR DESCRIPTION
**Pull request by matias9927**
Although JSR bytecodes cannot be generated by javac anymore, a classfile generated with a tool like JASM can still contain this bytecode. Should a program end with a JSR, there will be undefined behavior since the bytecode reads the address of the next instruction. In the case of Hotspot, this leads to a crash when generating oop maps. This fixes the calculation of basic blocks. 

The early exploration of this issue was done by @eme64 who also generated a reproducer.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335664](https://bugs.openjdk.org/browse/JDK-8335664): Parsing jsr broken: assert(bci&gt;= 0 &amp;&amp; bci &lt; c-&gt;method()-&gt;code_size()) failed: index out of bounds (**Bug** - P4)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Contributors
 * Emanuel Peter `<epeter@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20645/head:pull/20645` \
`$ git checkout pull/20645`

Update a local copy of the PR: \
`$ git checkout pull/20645` \
`$ git pull https://git.openjdk.org/jdk.git pull/20645/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20645`

View PR using the GUI difftool: \
`$ git pr show -t 20645`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20645.diff">https://git.openjdk.org/jdk/pull/20645.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20645#issuecomment-2298991196)

### CODE REVIEW

**Comment by dean-long** [comment]
Please add the regression test.

**Comment by dean-long** [comment]
What happens if the `jsr` is followed by illegal bytecodes that are unreachable, because the `jsr` has no `ret`?  Does GenerateOopMap still try to process the unreachable illegal bytecodes?

File: test/hotspot/jtreg/runtime/interpreter/LastJsr.jasm [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+super public class LastJsr
+{
+    public static Method test:"()V"
+    stack 100 locals 100
+    {
+        return;
+    LABEL:
+        nop;
```
Comment (dean-long) [inline test/hotspot/jtreg/runtime/interpreter/LastJsr.jasm:32]: Are these NOP instructions in both tests necessary?

Comment (matias9927) [reply test/hotspot/jtreg/runtime/interpreter/LastJsr.jasm:32]: I don't believe so, no. They were included as part of @eme64's reproducer.

**Comment by dean-long** [comment]
Let's say the JSR is not the last instruction, but it doesn't return because there is no RET instruction.  What is the effect of creating a basic block that won't be used -- is it harmless?  Do we need a test for that?

**Comment by matias9927** [comment]
Basic blocks have a property called `is_reachable` which seems to deal with unused basic blocks. The changes in my patch stop the new basic block from being marked or set to reachable. I think the current change is safe and unreachable basic blocks are already handled.

**Comment by matias9927** [comment]
To answer your question, I don't think that case needs to be tested.

**Review by dean-long — APPROVED** [review]
Looks good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. You might want to add @eme64 as co-contributor to this PR since he contributed the test.

**Comment by matias9927** [comment]
Thanks for the reviews @dean-long and @TobiHartmann! And thank you @eme64 for the reproducer!
/integrate

**Pull request closed without merging**